nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFiAtlasClient.java

```diff
@@ -0,0 +1,537 @@
+                if (urlMatcher.matches()) {
+                    flowPath.setGroupId(urlMatcher.group(1));
+                }
+            }
+            flowPath.setExEntity(flowPathEntity);
+            flowPath.setName(toStr(flowPathEntity.getAttribute(ATTR_NAME)));
+            flowPath.getInputs().addAll(toQualifiedNameIds(toAtlasObjectIds(flowPathEntity.getAttribute(ATTR_INPUTS))).keySet());
+            flowPath.getOutputs().addAll(toQualifiedNameIds(toAtlasObjectIds(flowPathEntity.getAttribute(ATTR_OUTPUTS))).keySet());
+            flowPath.startTrackingChanges(nifiFlow);
+
+            flowPaths.put(flowPath.getId(), flowPath);
+        }
+
+        nifiFlow.startTrackingChanges();
+        return nifiFlow;
+    }
+
+    @SuppressWarnings("unchecked")
+    private List<AtlasObjectId> toAtlasObjectIds(Object _references) {
+        if (_references == null) {
+            return Collections.emptyList();
+        }
+        List<Map<String, Object>> references = (List<Map<String, Object>>) _references;
+        return references.stream()
+                .map(ref -> new AtlasObjectId(toStr(ref.get(ATTR_GUID)), toStr(ref.get(ATTR_TYPENAME)), ref))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * <p>AtlasObjectIds returned from Atlas have GUID, but do not have qualifiedName, while ones created by the reporting task
+     * do not have GUID, but qualifiedName. AtlasObjectId.equals returns false for this combination.
+     * In order to match ids correctly, this method converts fetches actual entities from ids to get qualifiedName attribute.</p>
+     *
+     * <p>Also, AtlasObjectIds returned from Atlas does not have entity state.
+     * If Atlas is configured to use soft-delete (default), deleted ids are still returned.
+     * Fetched entities are used to determine whether an AtlasObjectId is still active or deleted.
+     * Deleted entities will not be included in the result of this method.
+     * </p>
+     * @param ids to convert
+     * @return AtlasObjectIds with qualifiedName
+     */
+    private Map<AtlasObjectId, AtlasEntity> toQualifiedNameIds(List<AtlasObjectId> ids) {
+        if (ids == null) {
+            return Collections.emptyMap();
+        }
+
+        return ids.stream().distinct().map(id -> {
+            try {
+                final AtlasEntity.AtlasEntityWithExtInfo entityExt = searchEntityDef(id);
+                final AtlasEntity entity = entityExt.getEntity();
+                if (AtlasEntity.Status.DELETED.equals(entity.getStatus())) {
+                    return null;
+                }
+                final Map<String, Object> uniqueAttrs = Collections.singletonMap(ATTR_QUALIFIED_NAME, entity.getAttribute(ATTR_QUALIFIED_NAME));
+                return new Tuple<>(new AtlasObjectId(id.getGuid(), id.getTypeName(), uniqueAttrs), entity);
+            } catch (AtlasServiceException e) {
+                return null;
```

Comment:
If an Exception is being thrown we should probably be at least logging it, rather than just swallowing it, no?

Reply:
Absolutely. Thanks!